[post_logout] fix state etc using form

cocoon opened the pull request:
After our discussion (https://github.com/sebadob/rauthy/discussions/682), going back and forth, validating everything again, I still come to the point where I think that the state field is not getting passed to the POST method.

Please verify again if this is wrong ^^

I think it is missed to passthrough at "LogoutHtml::build", the state is available before that in the "logout_request.state".


I now tried to fix the code (and it is working like expected now) to add it, but adding it to the file:
`\frontend\src\app.html`

required to add it to all templates, too, but I don't think it would be required everywhere.
In my first test I manually added it only to the logout.html that was already generated.

What would be a better way to only add it to the logout.html?

sebadob commented:
The state is already added.

This code block builds the location header if a `post_logout_redirect_uri` exists:

https://github.com/sebadob/rauthy/blob/a92374eed0fe082cd328cb4b2a54c6e8b9677357/src/api/src/oidc.rs#L675

But there is actually a tiny bug in the UI I just saw when looking into this.
The UI is not using the location header, but only the redirect URI.
This means the fix can be done in even a single line.

Everything that should be needed is something like

```js
window.location.href = res.headers.get('location');
```

in https://github.com/sebadob/rauthy/blob/a92374eed0fe082cd328cb4b2a54c6e8b9677357/frontend/src/routes/oidc/logout/%2Bpage.svelte#L51

Please have a look if this fixes your issue.

cocoon commented:
OK learned, I can debug the "res" object there with:

`$: res, console.log(res);`

and in my case there is no location:

![2025-01-08-001](https://github.com/user-attachments/assets/70591a4c-b930-4a84-8a9f-5cb49fc66852)

I will try to find out why not ...

cocoon commented:
Up to that line, the req object contains all data:

https://github.com/sebadob/rauthy/blob/a92374eed0fe082cd328cb4b2a54c6e8b9677357/frontend/src/routes/oidc/logout/%2Bpage.svelte#L33

So what also was working for me as a quick'n'dirty test is, to save the state and append it at the location url here:

 
```
 let mstate = ""; <------------

    onMount(async () => {
        const params = getQueryParams();
        postLogoutUri = params.post_logout_redirect_uri;
        const token = params.id_token_hint;
        const state = params.state;
        mstate = params.state; <--------------
...
async function handleRes(res) {
...
            var loc = postLogoutUri + "?state=" + mstate;
            alert(loc);
            window.location.href = loc;

```



![2025-01-08-002](https://github.com/user-attachments/assets/782b73fb-73bf-4f84-8115-ea8270ce1a07)


In this screenshot the following objects are printed:

```
onMount(async () => {
        const params = getQueryParams();
        postLogoutUri = params.post_logout_redirect_uri;
        const token = params.id_token_hint;
        const state = params.state;
        
        $: state, console.log(state);

        const csrf = window.document.getElementsByName('rauthy-csrf-token')[0].id
        saveCsrfToken(csrf);

        const immediateLogout = window.document.getElementsByName('rauthy-action')[0].id
        if ('true' === immediateLogout) {
            isLoading = true;

            const req = {
                id_token_hint: token,
                post_logout_redirect_uri: postLogoutUri,
                state: state,
            };

            $: req, console.log(req);

            let res = await logout(req);
            
            $: res, console.log(res);
            
            await handleRes(res);
        }
```

sebadob commented:
Pass in the correct `req` data object when doing the manual logout, in the same way as for the automatic one. The backend will only build and append the location header if a `post_logout_redirect_uri` is given. This is less error prone in the future than building the redirect  URI manually again, but would work as well in this case. This UI will receive a major rework anyway, as this is still Svelte 4 syntax.

cocoon commented:
Found something else here:

https://github.com/sebadob/rauthy/blob/a92374eed0fe082cd328cb4b2a54c6e8b9677357/src/api/src/oidc.rs#L659

I replaced it with 
`Json(params): Json<LogoutRequest>,`

and changed:
https://github.com/sebadob/rauthy/blob/a92374eed0fe082cd328cb4b2a54c6e8b9677357/frontend/src/utils/dataFetching.js#L93

with:
`body: JSON.stringify(data),`

Then I get the data in the POST method,

but end here:

`The page’s settings blocked the loading of a resource (connect-src) at https://win10pro.lab:5001/signout-callback-oidc?state=CfDJ8... because it violates the following directive: “connect-src 'self'”`

sebadob commented:
> I replaced it with
Json(params): Json<LogoutRequest>,

You must not do this. In fact, the `Query` is wrong as well, it should be `Form` by the spec, that's actually a bug. With it being `Query`, it will never accept the body properly. I also just saw that the `POST /logout` may need a bit bigger rework, because after another look at the RFC, a RP should be able to use the POST endpoint directly, as long as it sends an `id_token_hint`, which is not possible right now.

The `ReqPrincipal` needs to be chaned to be optional and then check inside the endpoint if either the principal is given or the `id_token_hint` exists. But you could also just take the `state` from the params for now and append it yourself, if it exists.

cocoon commented:
Great, looks like there is some light.

OK, will try the FORM.

And just one more thing, the Cookie "COOKIE_SESSION_FED_CM" needs to be flagged as secure now.

> Cookie rejected because it has the “SameSite=None” attribute but is missing the “secure” attribute.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie

sebadob commented:
> Great, looks like there is some light.

Yeah, I actually missed it from the RFC that it should work with a direct POST from an RP as well.

> And just one more thing, the Cookie "COOKIE_SESSION_FED_CM" needs to be flagged as secure now.

Yeah, the `SameSite=None` was a major concern from my side, when I implemented it, but they didn't care so far.

sebadob commented:
fyi: https://openid.net/specs/openid-connect-rpinitiated-1_0.html#RPLogout

sebadob commented:
> and that results in a error 415 Unsupported Media Type.

```js
export async function logout(data) {
    let body = new FormData();
    for (let key in data) {
        body.append(key, data[key]);
    }

    return await fetch('/auth/v1/oidc/logout', {
        method: 'POST',
        headers: getCsrfHeaders(),
        body,
    });
}
```

cocoon commented:
You had to post it quickly, right? So I had to compile again ^^
But still fails with the content error.

![2025-01-08-004](https://github.com/user-attachments/assets/bcb92b0d-ca90-495e-b5d1-53978e8d5574)

cocoon commented:
**Progress**

Found the issue from yesterday:

`let body = new FormData();`
creates multipart Form and
`getCsrfHeaders()`
sets 
`'Content-Type': 'application/json'`

I reverted my changes in this PR and replaced wit with some new changes just for easier reading:
`let body = new URLSearchParams();`
`getCsrfHeadersForm()`

Now it can extract the Form data and sets a location with state:

![2025-01-09-003](https://github.com/user-attachments/assets/00254733-fab6-419c-89c5-d00b0878a14a)


But I get stuck here:

![2025-01-09-004](https://github.com/user-attachments/assets/8be20690-bfe1-4a8a-85a9-9411cab6f515)

cocoon commented:
Found out more.

The post_logout uses MOVED_PERMANENTLY if it got a state, setting a LOCATION in headers:
```
return Ok(HttpResponse::build(StatusCode::MOVED_PERMANENTLY)
.append_header((header::LOCATION, loc))
```

https://github.com/sebadob/rauthy/blob/3b8cecbd46650a8aca71172863940384f1bb11d5/src/api/src/oidc.rs#L686

MOVED_PERMANENTLY = 301


The JavaScript part you pointed me to here:
https://github.com/sebadob/rauthy/pull/690#issuecomment-2577331068

where you thought a change like this could fix it:
`window.location.href = res.headers.get('location');`

will not work because the check "if response.ok is true" is not because 301 is not in the range for OK:

https://github.com/sebadob/rauthy/blob/a92374eed0fe082cd328cb4b2a54c6e8b9677357/frontend/src/routes/oidc/logout/%2Bpage.svelte#L51

> https://developer.mozilla.org/en-US/docs/Web/API/Response

> 
> Response.ok Read only
> 
>     A boolean indicating whether the response was successful (status in the range 200 – 299) or not.


And that is correct I think, because if a location header is set it should redirect on it's own and would not need the java script "window.location.href "?

But using the location header to redirect to another URL seems to be blocked by policy as shown.
This is how it looks in Chrome:

![2025-01-10-002](https://github.com/user-attachments/assets/74165846-0051-426e-8bb6-8c40adef7705)


I have seen it does an HTTP OPTIONS request then:

![2025-01-10-001](https://github.com/user-attachments/assets/ddbcb577-1c05-48fd-96d7-7715bc1f15f3)


Instead of a LOCATION header, i set it as a header named "loc" and used this in the javascript to set window.location.href to it.
That would work.

I pushed the changes as example.
But using MOVED_PERMANENTLY without LOCATION is not so nice, but working.

sebadob commented:
Instead of `301`, it makes sense to change it to a `200` to make it useable by RPs directly. These will probably expect an answer in the `2xx` range anyway. The `res.ok` check in the UI to extract the location is unnecessary, because a `301` would autoamtically redirect without the need for a manual redirection.

Just return a `200` and in the UI, check for an existing `location`. If it exists, redirect to it, and to `/auth/v1` otherwise. It makese sense to use location replace as well, because you can't go "back", since the logout is a one-time operation.

cocoon commented:
> Just return a `200` and in the UI, check for an existing `location`

Yes just thought the same, changed it now.

cocoon commented:
> use location replace as well, because you can't go "back", since the logout is a one-time operation

works 👍

sebadob commented:
> works 👍

If you remove the `WIP` and resolve the git conflicts, I can merge it.

cocoon commented:
> > works 👍
> 
> If you remove the `WIP` and resolve the git conflicts, I can merge it.

done.

You have seen I had to add another header to use form? It is a bit duplicated now.

```
function getCsrfHeadersForm() {
    return {
        ...HEADERS.form,
        'csrf-token': getCsrfToken(),
    }
}
```

sebadob commented:
> You have seen I had to add another header to use form? It is a bit duplicated now.

Yes I did. No worries though, the CSRF token will be removed anyway, when I rework the `/logout` endpoint that it will work with a direct `POST` from the RP in the future. I created an issue for this with #692 .

But this PR still has git conflicts, which is why I can't merge it. You should resolve these locally and push a merge from main for instance.

cocoon commented:
> But this PR still has git conflicts, which is why I can't merge it. You should resolve these locally and push a merge from main for instance.

🙈 resolved the conflicts, marked them but did not commit.
